### PR TITLE
Prohibit use of WorkflowClient from workflow code

### LINKED ISSUE #402: Prohibit use of WorkflowClient and Activity interfaces in workflow code.
**Is your feature request related to a problem? Please describe.**
New users might not be aware of nondeterministic requirements and might use WorkflowClient and Activity interfaces from the workflow code.

**Describe the solution you'd like**
Add explicit checks that throw exceptions from WorkflowClient and Activity APIs if they are called from workflow threads.

### PR DESCRIPTION
**Pull request by vkoby**
## What was changed
Added explicit checks that throw exception from WorkflowClient and Activity APIs if they are called from workflow threads.

## Why?
New users might not be aware of nondeterministic requirements and might use WorkflowClient and Activity interfaces from the workflow code.

## Checklist
1. Closes #402 

2. How was this tested:
Unit test

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowClientInternal.java [hunk 1]
```diff
@@ -414,4 +427,10 @@ public static WorkflowExecution start(Functions.Proc workflow) {
       A6 arg6) {
     return execute(() -> workflow.apply(arg1, arg2, arg3, arg4, arg5, arg6));
   }
+
+  private static void checkThread() {
+    if (Thread.currentThread().getName().startsWith("workflow")) {
```
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowClientInternal.java:432]: This doesn't look like a good method to check the thread for being a workflow. What if some user code names it's own thread "workflow"?

Looking at currently existing code, the right way to do this is
```
DeterministicRunnerImpl.currentThreadInternalIfPresent().isPresent();
```

I would actually add a method like
`WorkflowInternal#isCurrentThreadWokflow` containing this exact code.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowClientInternal.java [hunk 2]
```diff
@@ -414,4 +427,10 @@ public static WorkflowExecution start(Functions.Proc workflow) {
       A6 arg6) {
     return execute(() -> workflow.apply(arg1, arg2, arg3, arg4, arg5, arg6));
   }
+
+  private static void checkThread() {
```
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowClientInternal.java:431]: Please name with something more descriptive. Like `enforceWorkflowThread`.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowClientInternal.java (hunk 1 above)
Comment (vkoby) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowClientInternal.java:432]: I'm confused why the above guarantees that this is a wf thread as opposed to just any other internal thread. 
Though I agree that relying on the name of the thread is not very robust. This is where I encountered this way of identifying threads:
[ExceptionInSignalTest.java#L69](https://github.com/temporalio/sdk-java/blob/ce9c1dc5f9105f8631092364f4022407c3ae4ec3/temporal-sdk/src/test/java/io/temporal/workflow/signalTests/ExceptionInSignalTest.java#L69)

**Comment by Spikhalskiy** [comment]
Implemented in #556

**Pull request closed without merging**